pkg/farm/farm.go

```diff
@@ -0,0 +1,495 @@
+			builderMutex.Lock()
+			defer builderMutex.Unlock()
+			farm.builders[entities.LocalFarmImageBuilderName] = engine
+			return nil
+		})
+	}
+	if builderError := builderGroup.Wait(); builderError != nil {
+		if err := builderError.ErrorOrNil(); err != nil {
+			return nil, err
+		}
+	}
+	if len(farm.builders) > 0 {
+		defer fmt.Printf("Farm %q ready\n", farm.name)
+		return farm, nil
+	}
+	return nil, errors.New("no builders configured")
+}
+
+func NewFarm(ctx context.Context, name string, storeOptions *storage.StoreOptions, local bool) (*Farm, error) {
+	// Get the destinations of the connections specified in the farm
+	destinations, err := getFarmDestinations(name)
+	if err != nil {
+		return nil, err
+	}
+
+	return newFarmWithBuilders(ctx, name, &destinations, storeOptions, local)
+}
+
+// Done performs any necessary end-of-process cleanup for the farm's members.
+func (f *Farm) Done(ctx context.Context) error {
+	return f.forEach(ctx, func(ctx context.Context, name string, engine entities.ImageEngine) (bool, error) {
+		engine.Shutdown(ctx)
+		return false, nil
+	})
+}
+
+// Status polls the connections in the farm and returns a map of their
+// individual status, along with an error if any are down or otherwise unreachable.
+func (f *Farm) Status(ctx context.Context) (map[string]error, error) {
+	status := make(map[string]error)
+	var (
+		statusMutex sync.Mutex
+		statusGroup multierror.Group
+	)
+	for _, engine := range f.builders {
+		engine := engine
+		statusGroup.Go(func() error {
+			logrus.Debugf("getting status of %q", engine.FarmName(ctx))
+			defer logrus.Debugf("got status of %q", engine.FarmName(ctx))
+			err := engine.FarmStatus(ctx)
+			statusMutex.Lock()
+			defer statusMutex.Unlock()
+			status[engine.FarmName(ctx)] = err
+			return err
+		})
+	}
+	statusError := statusGroup.Wait()
+
+	return status, statusError.ErrorOrNil()
+}
+
+// forEach runs the called function once for every node in the farm and
+// collects their results, continuing until it finishes visiting every node or
+// a function call returns true as its first return value.
+func (f *Farm) forEach(ctx context.Context, fn func(context.Context, string, entities.ImageEngine) (bool, error)) error {
+	var merr *multierror.Error
+	for name, engine := range f.builders {
+		stop, err := fn(ctx, name, engine)
+		if err != nil {
+			merr = multierror.Append(merr, fmt.Errorf("%s: %w", engine.FarmName(ctx), err))
+		}
+		if stop {
+			break
+		}
+	}
+
+	return merr.ErrorOrNil()
+}
+
+// NativePlatforms returns a list of the set of platforms for which the farm
+// can build images natively.
+func (f *Farm) NativePlatforms(ctx context.Context) ([]string, error) {
+	nativeMap := make(map[string]struct{})
+	platforms := []string{}
+	var (
+		nativeMutex sync.Mutex
+		nativeGroup multierror.Group
+	)
+	for _, engine := range f.builders {
+		engine := engine
+		nativeGroup.Go(func() error {
+			logrus.Debugf("getting native platform of %q", engine.FarmName(ctx))
```

Comment:
Won't this debug message will look the same for every builder in the same farm?

Reply:
no, that actually returns the node name. I updated the fucntions to be `FarmNodeName`, `FarmNodeInspect`, etc to be clearer.